### DISCUSSION TITLE
Kotlin has a great features, let's stole 2 of them

### DISCUSSION THREAD

**Discussion by kefyru** [dotnet/csharplang#5974] [Language Ideas]
#1 The greatest  - lambdas and 'it' keyword

c#: Action<int> f = i => Console.WriteLine(i);
new: Action<int> f = { Console.WriteLine(it) };  // implicit ' it => ' 

c#:  list.Where(item => item == 33 );
new: list.Where { it == 33 };

c#: SomeFun(arg1, v => v * 2);
new: SomeFun(arg1) { it * 2 }; // if last arg is lambda

c#: DoSomething((arg1, arg2) => { ... } );
new: DoSomething { arg1, arg2 => ... };

c#: list.Where(i => i is not null)
           .Where(i => someList.Contains(i))
           .Select(i => convert(i))
           ...
new: list.Where { it is not null }
              .Where { someList.Contains(it) }
              .Select { convert(it) }

#2 infix functions
c#: var kv = new KeyValuePair<int,int>(2,3);
var d = new Dictionary<int,int>() {
   [1] = 11,
   [2] = 22,
   ...
}

new: 
var kv = 2 to 3;
var list = MakeDictionary(
   1 to 11,
   2 to 22, 
   ...
);

static class KeyValExt {
   // infix can be used only for static functions with two arguments 
   static infix KeyValuePair<K,V> to<K,V>( K key, V val) => new KeyValuePair<K,V>(key,val); 
   static Dictionary<K,V> MakeDictionary<K,V>(params KeyValuePair<K,V>[]) => ...
}

**Comment by kefyru** [comment]
i'm sorry ))
#1 The greatest - lambdas and 'it' keyword

```c#: Action<unt> f = i => Console.WriteLine(i);```
```new: Action f = { Console.WriteLine(it) }; // implicit ' it => '```

```c#: list.Where(item => item == 33 );```
```new: list.Where { it == 33 };```

```c#: SomeFun(arg1, v => v * 2);```
```new: SomeFun(arg1) { it * 2 }; // if last arg is lambda```

```c#: DoSomething((arg1, arg2) => { ... } );```
```new: DoSomething { arg1, arg2 => ... };```

```c#: 
list
   .Where(i => i is not null)
   .Where(i => someList.Contains(i))
   .Select(i => convert(i))```
...
```

```new: 
list
   .Where { it is not null }
   .Where { someList.Contains(it) }
   .Select { convert(it) }
```
#2 infix functions
```c#: 
var kv = new KeyValuePair<int,int>(2,3);
var d = new Dictionary<int,int>() {
   [1] = 11,
   [2] = 22,
   ...
}
```

```new:
var kv = 2 to 3;
var list = MakeDictionary(
   1 to 11,
   2 to 22,
   ...
);

static class KeyValExt {
   // infix can be used only for static functions with two arguments
   static infix KeyValuePair<K,V> to<K,V>( K key, V val) => new KeyValuePair<K,V>(key,val);
   static Dictionary<K,V> MakeDictionary<K,V>(params KeyValuePair<K,V>[]) => ...
}
```

**Reply by jmarolf** [reply]
While infix functions are not the same as [extension methods](https://docs.microsoft.com/dotnet/csharp/programming-guide/classes-and-structs/extension-methods). I don't think they are different enough to justify adding them:

```csharp
static class KeyValExt {
	static KeyValuePair<K,V> to<K,V>(this K key, V val) => new KeyValuePair<K,V>(key,val);
}
```

```csharp
var kv = 2.to(3);
var list = MakeDictionary(
   1.to(11),
   2.to(22),
   ...
);
```

**Reply by jmarolf** [reply]
As for the `it` keyword, I would add to the discussion here: https://github.com/dotnet/csharplang/discussions/91

**Reply by kefyru** [reply]
dots, braces, e.t.c => EVIL

**Reply by kefyru** [reply]
``` "Hello".append("Happy").append("World");```
or
``` "Hello" append "Happy" append "World";```
or 
``` "Hello" a "Happy" a "World";```

where
```
   static infix StringBuilder a(string left, string right) => new StringBuilder().Append(left).Append(right);
   static infix StringBuilder a(StringBuilder left, string right) => left.Append(right);
```
it's like operators, but extensible

**Reply by kefyru** [reply]
real case:
```
b.a("public sealed class ").a(id).a("_").a(m.Name).a("_").a(mId).a(m.TypeParameters)a(": IRpcMethod");
```

to
```
b a "public sealed class " a id a "_" a m.Name  a "_" a mId a m.TypeParameters a ": IRpcMethod";
```

**Reply by HaloFour** [reply]
As someone who has to regularly work with scalatest I vehemently disagree, those kinds of language-embedded DSLs are very difficult to read and involve having to effectively learn a new language within a language.  They almost always have weird warts, too, given that English grammar cannot be replicated using the combinations of these features.

**Reply by CyrusNajmabadi** [reply]
Neither of those "real cases" are comprehensible to me. If I can't figure out what they mean, they're not good examples to justify a language feature.

**Reply by kefyru** [reply]
It is a usual case in source generators, or building xml, json e.t.c...

**Reply by kefyru** [reply]
It's just build string from many small strings

**Reply by kefyru** [reply]
Enother real case (Kotlin)

```Kotlin
 val objectMap = MyMap.build { thisMap ->
        mapOf(
            "uuid<>" to meta { it.t.uuid = it.x.value },
            "InternalInfo" to IGNORE,
            "Properties" to mapOf(
                "Name<>" to meta { it.t.name = it.x.value },
                "Type" to createTypeSet(typeMap),
                "Use<>" to mapOf(
                    "ForItem" to addChild { "Use" attr "item" },
                    "ForFolder" to addChild { "Use" attr "folder" },
                    "ForFolderAndItem" to addChild { "Use" attr "both" },
                    "**" to context {
                        notify("unknown usage ${it.x.value}")
                    }
                ),
                arrayOf(
                    "Numerator",
                    "NumberType",
                    "NumberLength",
                    "NumberAllowedLength",
                    "NumberPeriodicity",
                    "CheckUnique",
                    "Autonumbering",
                    "CodeLength",
                    "DescriptionLength",
                    "CodeType",
                    "CodeAllowedLength",
                    "CodeSeries",
                    "InformationRegisterPeriodicity",
                    "WriteMode",
                    "RegisterType"
                ) to mapOf(
                    "<>" to addTemporary { it.x.name attr it.x.value }
                ),
                "Owners" to createStdMetadata(
                    "Attribute",
                    "Owner",
                    "_Owner",
                    MyMap.build(
                        "xr:Item<>" to addChild { CMetaTypeReference(it.x.value) }
                    )
                ),

                "RegisterRecords" to createStdMetadata(
                    "Std",
                    "RecordTypes",
                    null,
                    MyMap.build(
                        "xr:Item<>" to addChild { CMetaTypeReference(it.x.value) }
                    )
                ),

                "Synonym.v8:item" to mapOf(
                    "v8:lang" to null,
                    "v8:content" to null
                ),
            ),
            "ChildObjects" to mapOf(
                V8XmlSchema.childKinds to createMetadata(thisMap, V8XmlSchema.kindMap)
            )
        )
    }

```

and this code without fixins:

```Kotlin
 var objectMap = MyMap.build { thisMap ->
        mapOf(
            "uuid<>".to( meta { it.t.uuid = it.x.value } ),
            "InternalInfo".to( IGNORE ),
            "Properties".to( mapOf(
                "Name<>".to( meta { it.t.name = it.x.value }),
                "Type".to( createTypeSet(typeMap)),
                "Use<>".to( mapOf(
                    "ForItem".to( addChild { "Use" attr "item" } ),
                    "ForFolder".to( addChild { "Use" attr "folder" }),
                    "ForFolderAndItem".to( addChild { "Use" attr "both" } ),
                    "**".to( context {
                        notify("unknown usage ${it.x.value}")
                    } )
                )),
               ......
            )
        )
     )

```

**Reply by kefyru** [reply]
braces === EVIL ))

**Reply by HaloFour** [reply]
@kefyru 

> braces === EVIL 

They're definitely not evil.  IMO, your examples with the braces are much easier to read.

**Reply by fabianoliver** [reply]
I could certainly see some cases where it makes things a little more readable.

```
cars select {it.driver} where {it.age > 50}
```

Is arguably a bit nicer than

```
cars.select(x => x.driver).where(x => x.age > 50)
```

It's probably the very reason why at some point, someone introduced special-purpose LINQ query syntax?
Having a means to achieve something similar for other use cases is probably fair to debate.

Having said that, while I don't have any stats, I've personally never seen anyone actually use query over method syntax. And I'd imagine having to deal with conflicts between this syntax and reserved keywords (current ones, as well as future ones), could be a bit of a nightmare. So I'm not quite convinced it would really be a net benefit.

**Reply by kefyru** [reply]
full example: Real mach easier??? Try to find unnecessary brace )))
```Kotlin
val objectMap = MyMap.build { thisMap ->
        mapOf(
            "uuid<>".to(meta { it.t.uuid = it.x.value }),
            "InternalInfo".to(IGNORE),
            "Properties".to(mapOf(
                "Name<>".to(meta { it.t.name = it.x.value }),
                "Type".to(createTypeSet(typeMap)),
                "Use<>".to(mapOf(
                    "ForItem".to(addChild { "Use".attr("item") }),
                    "ForFolder".to(addChild { "Use".attr("folder") }),
                    "ForFolderAndItem".to(addChild { "Use".attr("both") }),
                    "**".to(context {
                        notify("unknown usage ${it.x.value}")
                    })
                )),
                arrayOf(
                    "Numerator",
                    "NumberType",
                    "NumberLength",
                    "NumberAllowedLength",
                    "NumberPeriodicity",
                    "CheckUnique",
                    "Autonumbering",
                    "CodeLength",
                    "DescriptionLength",
                    "CodeType",
                    "CodeAllowedLength",
                    "CodeSeries",
                    "InformationRegisterPeriodicity",
                    "WriteMode",
                    "RegisterType"
                ).to(mapOf(
                    "<>".to(addTemporary { it.x.name.attr(it.x.value })
                )),
                "Owners".to(createStdMetadata(
                    "Attribute",
                    "Owner",
                    "_Owner",
                    MyMap.build(
                        "xr:Item<>".to(addChild { CMetaTypeReference(it.x.value) })
                    )
                )),

                "RegisterRecords".to(createStdMetadata(
                    "Std",
                    "RecordTypes",
                    null,
                    MyMap.build(
                        "xr:Item<>".to(addChild { CMetaTypeReference(it.x.value) })
                    )
                ))),

                "Synonym.v8:item".to(mapOf(
                    "v8:lang".to(null),
                    "v8:content".to(null)
                ))),
            ))),
            "ChildObjects".to(mapOf(
                V8XmlSchema.childKinds.to(createMetadata(thisMap), V8XmlSchema.kindMap)
            ))
        )
    }
```

**Reply by HaloFour** [reply]
@kefyru 

> Try to find unnecessary brace

At least you can count braces.  Without them you have to rely on mentally parsing how the compiler will interpret the whilespace around each token.  Is that an argument?  A method call?  Is there a dot there or braces?  It requires the human reader to be intimately aware of how the machine will end up inserting that punctuation.

**Reply by kefyru** [reply]
it is the same as
``` 1 * (2 + 3/5) / 4```
replace with
``` 1.*(2.+(3./(5)))./(4)```

**Reply by kefyru** [reply]
i  refuse to understand you )))

**Reply by HaloFour** [reply]
No, it's not the same as that.

**Reply by kefyru** [reply]
for me, after trying Kotlin, i cant do the same things in c#. But i love c# & dotNet and i want it fallow modern trends

**Reply by HaloFour** [reply]
You can do the same things, just with different syntax.  That would be true regardless of which other languages you use.  There are plenty of great features in Kotlin, many which should be considered for C#, but adopting them to C# syntax and idioms.

**Reply by CyrusNajmabadi** [reply]
> It's just build string from many small strings

In C# we would just use `+` for this.  :)

**Reply by CyrusNajmabadi** [reply]
Let's step back and not get into a violent argument about syntactic preferences :)

**Reply by kefyru** [reply]
> > It's just build string from many small strings
> 
> In C# we would just use `+` for this. :)

No we cant. + does not build something (for example string builder) it just add or concat. If you about overload - try to overload say string class))

**Reply by CyrusNajmabadi** [reply]
`+` can be overloaded however a type wants to overload it.  

You also mentioned: `It's just build string from many small strings`, which is precisely what it can do.

**Comment by CyrusNajmabadi** [comment]
@kefyru First, thanks for the feedback.  Definitely great to hear about hte things you like in othr languages which you would like to see in C#.  Second, let me say that we're always considering syntactic niceties and improvements.  However, it's not a simple thing and we are very cautious about just adding ways to do things that are already possible just with an entirely different syntax.  We get enough feedback from people not happy that there are so many ways to do things already, and these syntactic/stylistic differences really add to that.

So, to that end, in order to make these sorts of changes we really want deep reasons why the new form is truly superior and a good story around how the two forms will coexist in the future.  Saying things like "braces === EVIL" (while likely tongue in cheek) don't really help here as this then becomes a discussion about how to make the language ideal for you, rather than how to design a language that will feel good for the millions of customers already using it as it exists today who will now have to absorb this new feature and use it among 20+ years of code investments already using different styles.

**Reply by kefyru** [reply]
And that s why we love c#)
And braces can be removed from code by setting for them color=background in ide))))

**Reply by TonyValenti** [reply]
When I read this, I immediately hated it, but the ```it``` idea kind of has grown on me, but with a slight modification.

Today, when you create a property, the setter has an invisible parameter, ```value``` , which contains the value that is being set.

I think I would like the ```it``` proposal if the variable was renamed to ```value```.

ie:
```
var v = MyList.Where(value != 23).FirstOrDefault();
```

**Reply by kefyru** [reply]
I still think that rather than reinventing the wheel, it's better to borrow a good idea from another language. Kotlin implemented a successful idea with `it`, it is very convenient and understandable. Using `value` will cause problems with the implementation of properties setters, where `value` is usually used as a parameter.

Besides, `Where(value==something)` is hard to understand - it's `Where<bool>` or `Where<Func<bool>>`. This is not clearly visible from the syntax. Therefore, it is also necessary to implement a new lambda construction as in kotlin: `Where( {it == something} )` or even better `Where { it == something }`

**Comment by jnm2** [comment]
The `it` discussion is https://github.com/dotnet/csharplang/discussions/602.

Similar: https://github.com/dotnet/csharplang/discussions/1151, https://github.com/dotnet/csharplang/discussions/3075
